### ISSUE TITLE
CoqIDE does not work on Windows

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#9885]
Since the switch to GTK3, CoqIDE has quite a few issues on Windows:

* It crashes on file open - this can be fixed by providing a compiled scheme file
* Icons don't work - this can be fixed by providing an icon scheme. This has the issue that the complete icon scheme is very large (40MB or so) and that coqtop scans the schare folder for plugins and adding the icons adds about 20.000 system calls to the start of each coqtop for the plugin scanning. If only those icons are copied which ara actually used, it is ok, but I have no idea if it always uses teh same icon size on all systems / settings.
* Syntac highlighing doesn't work
* Even with the file open patch mentioned above, CoqIDE usually crashes on file open. I coudln't find out why - need to debug it.

Otherwise it seems to work - if one can get a fil opened, one can get step through it.

**Comment by vbgl** [comment]
For the record: these issues are not specific to Windows. The macOS version that is built and packaged on azure has similar behavior.

**Comment by MSoegtropIMC** [comment]
It is still on my todo list ...
If this is of interest I could do a PR for the fixes I already have, although I am not so sure about the icon set to use.

**Comment by ppedrot** [comment]
Do not hesitate to submit partial fix PRs, a half-broken CoqIDE is better than a completely broken one!

**Comment by MSoegtropIMC** [comment]
Since the fixes I mentioned are all about system specific installation stuff (creating and copying around some files), this would have to be done separately for Windows and Mac (most likely at least). Who would be the expert for Mac to ask? Or should I just provide a patch for Windows and I help then whoever maintains the Mac build to port it?

**Comment by vbgl** [comment]
I fear that there is no macOS expert around… I’ll do my best to help, though.

**Comment by newca12** [comment]
FWIW I do not reproduce any of the problems mentioned here with coq master HEAD manually compiled on macOS 10.13.6 but I see them all if I use : https://github.com/coq/coq/pull/9891#issuecomment-480276966

**Comment by ejgallego** [comment]
Thanks @newca12 , do you see any useful log info? Likely a problem with paths?

**Comment by newca12** [comment]
Providing gschemas.compiled and icons/Adwaita (60MB) seem to fix all issues on macOS with the Azure artifact

**Comment by MSoegtropIMC** [comment]
@newca12: thanks, interesting. Last time I tried this on Windows (2 1/2 weeks back) after providing the schema file and the icons I still had the issue that CoqIDE crashed on file open fairly regularly, especially after changing folders. But this might be a Windows specific "current path" issue.

Can you check what happens if you open say 10 different .v files from 10 different folders?

**Comment by MSoegtropIMC** [comment]
Btw.: Is there any good reason why coqtop scans the share folder for plugin files (OCaml shared libraries?). Does it ever happen that plugin files are actually in there? And even if so, it might make sense to reverse the search order (search the lib folder first). I guess on Linux the share folders can get really huge and scanning the complete folder for libraries doesn't seem to be a very sensible thing to do. But maybe this is a Windows specific behavior of OCaml and nothing explicitly coded in Coq.

**Comment by newca12** [comment]
I confirm that there does not seem to be any problem with file open on macOS after providing the schema file and the icons.

**Comment by vbgl** [comment]
@newca12 Can you please provide some guidance on how to “provide the schema files and the icons”? Your help is appreciated.

**Comment by MSoegtropIMC** [comment]
@vbgl : 
The creation of the schema file I explained in #9279 :

Apparently there is a file missing, which can be created (in the windows case) via:
```
glib-compile-schemas /usr/x86_64-w64-mingw32/sys-root/mingw/share/glib-2.0/schemas/
```
from the build cygwin console. I guess it is similar for Mac. If a Mac is setup like Linux with globally shared folders, this should be it.  Otherwise the resulting gschemas.compiled file needs to be copied to
```
<coqinstallroot>\share\glib-2.0\schemas
```
For the icons you need to install the Adwaita icon scheme via whatever package manager you have on Mac. Again, if there is something like a Coq install folder, you need to copy them to
```
<coqinstallroot>/share/icons/Adwaita
```
The main downside of this is that it makes Coq much slower, at least on Windows, because it scans this folder regularly for cmxs files.

**Comment by newca12** [comment]
@vbgl the only file that is really needed to be compiled on macOS to fix the open file issue is org.gtk.Settings.FileChooser.gschema.xml from gtk+3 package. So you can glib-compile-schemas a directory that contain only that file to produce gschemas.compiled.  Regarding the icons I haven't notice the slowdown but I may not have been vigilant enough.

**Comment by vbgl** [comment]
Thank you for this detailed help!

**Comment by MSoegtropIMC** [comment]
> Regarding the icons I haven't notice the slowdown but I may not have been vigilant enough.

It might well be that this is a Windows specific problem. I would think that Mac has like Linux some sort of cache for knowing the locations of executable files. Windows is silly enough to search half of the hard disk each time it is looking for an executable.

**Comment by vbgl** [comment]
Will we be able to provide an installer for Windows of a working version of Coq 8.10+β1?

I wished we could tag that version tomorrow (April 30th) and announce it the day after (May 2nd). Thank you.

**Comment by MSoegtropIMC** [comment]
@vbgl : I think I can make it until next Monday but not before. Unfortunately this was really a lot of work so far:

- Issues cannot be debugged by behavioral analysis (with tools like ProcMon which record system calls), so one needs to debug it
- Bytecode builds of Coq on Windows are tricky with the old make based build system => switch to dune
- In the dune build (not in the make build) not only coqide but also coqidetop has issues. It is rather tricky to run coqidetop in a debugger, because it is usually started by coqide and given the communication ports. So either one has to debug this with adding traces or by implementing in CoqIDE a method to connect to a pre started coqidetop (see discussions with @gares).
- This leads to that one finally needs a proper development environment on Windows - messing around with Coq without Merlin and debuggable byte code builds is no fun.
- I picked up my old efforts to come up with a script which properly sets up a reasonable development environment on Windows. Doing so I investigated the recent cygwin cross opam for MinGW <https://fdopen.github.io/opam-repository-mingw/installation/>. While cygwin/MinGW cross opam was really far away from being able to build CoqIDE two or even one years back, this didn't look bad, so I gave it a try. I actually managed with just a handful of changes to the repo to build CoqIDE for MinGW using opam yesterday (see <https://github.com/MSoegtropIMC/opam-repository-mingw/commit/4dd5d3e67154c6f667fd96f062916eac9b3e89c7>). Also merlin and tuareg seem to work with some patches, but I didn't try hard stuff as yet. The required repo changes have been a bit tricky for a opam newcomer like me, e.g. it took me a while to figure out how to incorporate the flexlink patches required to build lablgtk3. Also this needs quite a bit of dependency regularization before it can be integrated upstream.
- I also spent quite some time on figuring out how to create and use debuggable byte code builds of coq executables using dune. This also works meanwhile but is not well tested as yet. Especially for GUI stuff this is a bit tricker than on Linux, because the Windows GUI apps require a specific environment in the file system to run.

So under the line chances are good that I can do some debugging of the issues in the next days, but I didn't start as yet and just worked on the environmental requirements up to now. But I made good progress there.

Btw.: Since the build via opam looks fairly reasonable, I plan to switch the build to opam and dune some time soon.

**Comment by Zimmi48** [comment]
I can see two solutions here:

1. Delay the beta release, hoping that the Windows problems will be fixed by then.
2. Do what @ejgallego suggested as a possible solution if too much trouble arose, by reverting to gtk2 *only* in `v8.10` and *only* in Windows. I have no idea how feasible this is in practice.

**Comment by ejgallego** [comment]
> 2\. Do what @ejgallego suggested as a possible solution if too much trouble arose, by reverting to gtk2 _only_ in `v8.10` and _only_ in Windows. I have no idea how feasible this is in practice.

We should indeed provide the gtk2 version anyways. A plan could be like:
1. create a repos on Coq for the gtk2 version
2. switch the windows build scripts to use that version

**Comment by MSoegtropIMC** [comment]
I think reverting to GTK2 would be quite a lot of effort and I don't see how you want to do it just for Windows. One would need two source code folders for Coqide then unless one wants to have a separate branch for Windows. I would think that this would likely take longer than until Monday.

@ejgallego : I tend to not agree that we should support 2 versions. We hardly have enough man power to support one.

I would say let's wait until Monday and see if it is fixed by then. Maybe I can even fix it until Thursday.

**Comment by ejgallego** [comment]
Compiling the GTK2 version for Windows is actually pretty easy; it is distributed as a separate package so all that you have to do is to compile and install Coq, then install the extra package.

**Comment by MSoegtropIMC** [comment]
> it is distributed as a separate package so all that you have to do is to compile and install Coq, then install the extra package.

I guess this is true for just building with opam + dune, but the scripts which create the installer are still make based and have a lot of complicated and partly messy stuff in them. It is not just about building this stuff, it is about creating an installer (and the open source package) from it. I would think that switching all the installer logic to opam and dune is at least 1 week full time of work, maybe more.

**Comment by gares** [comment]
>     * n the dune build (not in the make build) not only coqide but also coqidetop has issues. It is rather tricky to run coqidetop in a debugger, because it is usually started by coqide and given the communication ports. So either one has to debug this with adding traces or by implementing in CoqIDE a method to connect to a pre started coqidetop (see discussions with @gares).

I was travelling and then I forgot to send you this piece of info (and I don't find the original thread, so I do it here).

From one shell
```
$ ocamldebug -s localhost:99993 bin/coqidetop.byte
	OCaml Debugger version 4.07.0

(ocd) set loadingmode manual
(ocd) step
Loading program... 
Waiting for connection...(the socket is localhost:99993)
```

From another shell
```
$ bin/coqide.byte -coqtop-flags extra-env=CAML_DEBUG_SOCKET=localhost:99992 -debug
```

Now from the first shell (I just do a few steps, you may set breakpoints and stuff)
```
done.
Time: 1 - pc: 8320 - module Stdlib
27     (Invalid_argument "index out of bounds")<|a|>
(ocd) break @ Coqtop 203
Breakpoint 1 at 4862660: file toplevel/coqtop.ml, line 203, characters 39-68
(ocd) run
```

I've just tried and for some reasons the breakpoint is not honoured ;-) but I've no more time on this.
Good luck

**Comment by vbgl** [comment]
Thank you Michael. We’ll wait a bit before announcing the β until CoqIDE works on Windows.

**Comment by MSoegtropIMC** [comment]
@vbgl & all: I just wanted to give a short progress report:

I have now a working OCaml development and debug environment on Windows. The main meanwhile solved working points have been:
- Switching to opam for the prerequisites - this required pushing some patches (e.g. the flexlink patch) to the MinGW opam (it is not yet in flexlink master but in MinGW opam master now) and enabling some additional builds (e.g. lablgtk3 did not work before).
- Switching the Coq build to dune (easy)
- Exploring Dune byte code builds (easy, I love it)
- Teaching cygwin emacs to understand Windows path. Cygwin does understand windows path, but emacs has its own path handling and didn't. This means that it is hard to use MinGW tools like ocamldebug or merlin in cygwin emcas. The MinGW emacs is not much of an option because it essentially installs another Posix layer (some 25.000 files) and doesn't understand cygwin paths then. There are solutions on the lisp level which work by path translation, but I really needed more native support. Unfortunately the corresponding emacs file is a bit like an archaeological excavation site: layers on layers on layers on layers and really nasty "don't do this or that here or the GC will kill you" stuff. Half way through i was so ***** that I started to rewrite it from scratch, but half way through that, when it came to the details of ancient OSes upstream emacs might still want to support, I thought quick and dirty might still be faster. What can I say - it works. 
- I also looked shortly into the alternative of setting up all the OCaml tools such that they are real cross tools (designed to run on cygwin but to create MinGW code) but this is rather tricky because many of the created executables are called in the build (say coqc) and the support for this in dune is still rudimentary. E.g. one would need at least 2 ocaml compilers, one to create tools like dune which shall run under cygwin and one to create tools like coqc which shall run under MinGW and eventually a third one which runs under MinGW for native compute.
- Some minor adoption of OCaml related emacs files (path and \r handling mostly)
- Some minor adoption of VSCode merlin handling.
- Automated setup of all of this (finished except for recompiling the patched emacs).
- Understanding how to run coqide and coqidetop in the debugger together.

All in all this are not so many changes, but for each change it takes quite some time to find the root cause and what to adjust how. Just the Emacs stuff really took a while to fix.

This was all finished around yesterday early afternoon and I started fixing the issues I see in the Dune build I don't see in the traditional build. Up to now this have been path issues which could be fixed by some creative prefix options in the configure and install stage. But I have still one issue during startup (coqtop died badly) which I only see with the dune build. Nothing blocking up to now, it just needs to be done. I have not that much time today, so it will likely be tomorrow. After this I can hopefully start to debug the issue I see in the traditional build, so honestly it might Thursday until it is all done.

Even though it causes delays, I am sure it is worthwhile. After all this work we will be in a much better position to maintain the Windows build in a professional way. Developing on Linux and relying on luck that it will also work on Windows is not a long term option.

**Comment by MSoegtropIMC** [comment]
An update:

I found the issue that I get a "Coqtop died badly" on the dune byte build but not with the legacy build. The issue is that the lablgtk hack I did years back is not used in coqide in master. It is this line in ide/coq.ml:

```
let gio_channel_of_descr_socket = ref Glib.Io.channel_of_descr
OR
let gio_channel_of_descr_socket = ref Glib.Io.channel_of_descr_socket
```

The point here is that on Linux file handles and socket handles share the same number space, while on Windows these are separate types and can have the same numbers. So one has to know if one has a file handle or a socket handle. I did a hack years back and added one function to lablgtk to solve this. Interestingly while the patch was ported to GTK3, the corresponding usage in ide/coq.ml vanished, as it looks long before the move to GTK3. I looked at quite a few versions of the file but couldn't find it. Maybe it is just luck that for whatever reason in the native build the socket and file handles are different while they collide in the byte code build. Anyway, if I use `Glib.Io.channel_of_descr_socket` in coq.ml it works. Possibly this is the reason why parallel / async proofs have been instable. Btw.: `Glib.Io.channel_of_descr_socket` is still used in master in coqide_WIN32.ml.in.

@ejgallego : can you explain what dune does with .in files like coqide_WIN32.ml.in? Maybe this is also part of the difference between dune and makefile builds.

@gares : In case you remember something about the history of the channel_of_descr_socket story, please share.

Summary: I can now start to debug the real issue seen in the native code build :-)

**Comment by MSoegtropIMC** [comment]
@ejgallego : 

I found the piece for the .in files in ide/dune (was grepping for WIN32). It reads:

```
(rule
 (targets coqide_os_specific.ml)
 (deps (:in-file coqide_X11.ml.in)) ; TODO support others
 (action (copy# %{in-file} %{targets})))
```
Any idea how I could do this properly?

**Comment by MSoegtropIMC** [comment]
@vbgl: my daily update on this

The issue is that the bugs / crashes of CoqIDE (usually it segfaults after a few opens) depend a lot on circumstances.
- Even after fixing the .in file stuff, the dune build still behaves differently than the traditional makefile build. Both have issues, but not the same ones.
- Interestingly the dune build has more issues when compiled with the WIN32 os dependent file than when compiled with the (empty) X11 os dependent file.
- I also investigated byte code builds using the traditional make files. The issues of the native executable can be reproduced quite exactly with the makefile bytecode builds, but not when run under ocamldebug. If run under ocamldebug, the coqide UI doesn't come up at all. Interestingly this does work with the dune bytecode builds - they run under ocamldebug.

Under the line I currently don't see a reasonably fast way to debug this with ocamldebug. I will fall back to adding tracing. Luckily merlin runs well in Emacs and VSCode.

Honestly I don't see a chance before Monday. The code is quite complicated and adding traces will take time. I would suggest we wait until Monday. If this is still not fixed by then, we should discuss alternatives.

**Comment by ejgallego** [comment]
@MSoegtropIMC please open an issue for the arch-specific file and I will fix it ASAP.

**Comment by vbgl** [comment]
Thank you Michael for the update. What alternative do you have in mind? I see two possibilities (if releasing a working CoqIDE is not feasible):

 -  tag the version 8.10β1 as-is, knowing that CoqIDE is hardly usable on Windows;
 -  release Coq without CoqIDE and separately release CoqIDE (hopefully we can find a working version in the recent past).

**Comment by gares** [comment]
> @gares : In case you remember something about the history of the channel_of_descr_socket story, please share.

When I crafted the old SDK for windows I had a patch for lablgtk2 (that I could not get merged upstream, I don't recall why) to export windows specific socket functions because as you say the unix ones don't work properly on windows (sockets). Indeed glib provide OS dependent APIs. You had this patch in your build script too IIRC, and I see that now it has been ported to lablgtk3 https://github.com/coq/coq/blob/master/dev/build/windows/patches_coq/lablgtk-3.0.beta5.patch
I guess you are still using it in your dune builds, right?

We should also try to push this patch upstream, since it is in no way Coq specific CC: @garrigue

**Comment by garrigue** [comment]
I see no problem in merging that patch, both in lablgtk2 and lablgtk3.
(I'm not doing much windows testing these days, but if you have done it it's ok).

BTW, it's a bit strange that lablgtk seemed to contain windows specific code that was not working. Or did it work just partially?

**Comment by gares** [comment]
IIRC on windows the w_sock handle is not an integer, so I guess a mapping behind the scenes is necessary, while on unix the mapping is just the identity function. Using the identity mapping on windows pipes was OK (I guess their handle is an integer), while was not working properly on sockets. 
I imagine the `win32` specific version of the first API is just an alias of the unix one (eg no mapping is necessary).

**Comment by MSoegtropIMC** [comment]
@garrigue : I would say that the patch is more a hack. It works for us but it requires changes to code which works on Linux so that it works on both Windows and Linux (one has to use different functions for socket handles and file handles). It might be fine if this is properly documented, but then this change also would have to be complete for all function which accept both file and socket handles, which is not the case. Right now we have separate functions just for those we need (that is one function).

So the options to get this upstream are:

- Use some form of abstract handle on teh APU user side and map these to whatever handles the OS uses (as @gares suggsted above). This has the disadvantage that such handles cannot be passed any more to posix functions. Of cause this can be fixed by providing conversion functions and it can be made type safe.
- Make this split between socket and file handles everywhere and document it properly.

I tend to the first option, but it might break a lot of code out there.

**Comment by MSoegtropIMC** [comment]
> What alternative do you have in mind?

Honestly I didn't think about it much. The plan is still to get it fixed asap. I would prefer to defer this discussion until Monday. In case I can't fix it until then, I will hopefully have some increased knowledge about what effort it will be to fix it. Depending on this we can decide to have a temporary "No CoqIDE beta" or we fallback to GTK-2 on Windows, which I would prefer to avoid.

**Comment by MSoegtropIMC** [comment]
@vbgl : my update on this:

Debugging is still rather slow on Windows. It is simply hard to know what to expect. E.g. the reason coqide does not bring up a UI if run under ocamldebug is that that it calls coqidetop early once for some sort of parameter analysis and this call returns an error. Debugging this is tricky because the redirection to an already running coqidetop does not work for this early call, only for the main call to coqidetop. The early call gets an uncaught exception very early (before it enters the main function) so bisection with traces is also tricky. Calling coqidetop in the debugger (or without it) with exactly the same arguments and (as far as I can tell) same environment mysteriously doesn't lead to an exception. Another things is that I couldn't bring printing backtraces on an uncaught exceptions to work. Even if I hack the OCaml library to always set caml_backtrace_active very early during startup, I don't get a backtrace. This is not because there wouldn't be debug info (the corresponding test succeeds), the log is simply empty. I would expect that in a byte code build with debug info it should be possible to at least show the location of where the exception was thrown.

I also tried to trace bisect the issues with file open, but this is also not easy with GUI apps, so I reverted to get the debugger to work.

All in all I would expect that this is 2..4 weeks more of work (doing it in parallel with my main work). So I would suggest to release the beta without a working CoqIDE. Reverting to GTK2 and back is probably not worth the effort for a beta and the time frame in question. We can discuss further steps if I still see no land after 2 weeks.

**Comment by vbgl** [comment]
Thanks for the update and thank you again for your hard work. Let’s launch the β-testing phase for Coq 8.10 and have this issue postponed.

**Comment by gares** [comment]
> calls coqidetop early once for some sort of parameter analysis

IIRC this piece of code is not particularly nice, so I would not mind if the quick fix was just to filter the command line arguments in a more direct way. Given the problem arises on dune build I suspect the binary does not find COQLIB (but I did not check, so I may be wrong in guessing).

**Comment by MSoegtropIMC** [comment]
> Given the problem arises on dune build I suspect the binary does not find COQLIB (but I did not check, so I may be wrong in guessing).

This happens on the traditional build. Also I fixed the dune build such that it does find COQLIB. In the dune build I build Coq like this:

    #Mimic with the -xxdir options what opam install with --prefix actually does
    ocaml.exe configure.ml -no-ask -native-compiler no -prefix "./" -libdir "./lib/coq" -datadir "./share/coq"
    make -f Makefile.dune voboot $MAKE_OPT
    sed -i 's|coqide_X11\.ml\.in|coqide_WIN32\.ml\.in|' ide/dune
    dune build $MAKE_OPT
    rm -rf "$PREFIXCOQ"
    mkdir -p "$PREFIXCOQ"
    dune install $MAKE_OPT --prefix="$PREFIXCOQ"

Essentially I do the same trick as in the makefile build, I configure Coq with relative paths and install it to absolute paths. I am not so sure what to do with -datadir, but the -libdir above has the desired effect. Interestingly this always worked regardless on what the current directory is - I guess there is some handling for this in the coq tools and one just has to reconstruct the same settings in dune as they used to be.

> so I would not mind if the quick fix was just to filter the command line arguments in a more direct way

From my point of view the major issue is not that there is a bug, but that I cannot debug this bug efficiently. I just see this is a good opportunity to get the development environment on Windows setup and tested properly. It is worthwhile to analyze and understand all the differences between dune and make builds, byte code and native builds and running an application with and without debugger. The goal is that there are no such differences. So in a sense it is good that I see these differences now and can analyse their root cause. At least as long as we are in a beta phase, I see this as a more important goal than getting CoqIDE fixed by hacks.

Of cause the situation will be different as soon as we approach a release. Then a more pragmatic approach might be called for. But I hope the issues are solved by then.

**Comment by maximedenes** [comment]
> Of cause the situation will be different as soon as we approach a release.

Well, we are already two weeks past the expected release date... So I was hoping we could find a quick fix, as shipping CoqIDE for gtk3 in a final version without it having been tested in a beta (on Windows, at least) is certainly not a desirable situation.

But don't get me wrong, I understand this is very tricky and you already put a lot of effort in it.

**Comment by MSoegtropIMC** [comment]
@maximedenes , @vbgl : 

What is the planned release date for 8.10.0? Indeed it would be good to have some headroom after fixing the bugs. before 8.10.0.

Just fixing the issue @gares mentioned wouldn't really fix CoqIDE. It just would allow me to get one step further in debugging it. This is just the first issue which keep CoqIDE from starting up when started with ocamldebug. If fixing this gets me anywhere close to debugging the real issue is a different question.

**Comment by MSoegtropIMC** [comment]
I made some progress on the debugging front. See:

<https://github.com/ocaml/ocaml/issues/8678>

I will do some quick and dirty hack to get around this issue and hope that I can make progress with the real issue then.

**Comment by MSoegtropIMC** [comment]
Latest updates: it proved rather tricky to even hack the issues in ocamldebug. I had to dig rather deep into the topic. I still don't see how this is intended to work. I will now introduce a global variable which stops an executable from talking to the debugger and simply clear this just in those executables I want to debug.

**Comment by MSoegtropIMC** [comment]
My weekly progress report on this:

- I am fairly ok with ocamldebug meanwhile. I fixed the issue with propagation of the debugger connection environment variable to child processes and can debug now with ocamldebug.
- What reliably does not work is the coqidetop started when opening a second file in another folder. This is then the 4th coqidetop process spawned (one for analyzing parameter, one with -batch, one for the first document and one for the second). So the trick suggested by Enrico does not work. What likely would work is to change the coqidetop settings in the preferences and give the parameters there after opening the first file. Just (how could it be otherwise) when I open one file, CoqIDE crashes when opening the preferences window in some C code. So I can't use ocamldebug to debug this, since it just looses connection on a seg fault.
- Loading and running the executable in gdb to debug the seg fault from C code works, but then it doesn't bring up the UI again. Process starting and environment look good ...

I guess I will simply implement a facility to trace instructions in the OCaml bytecode interpreter (something like this exists, but it goes to stdout and cannot be easily enabled) and analyze the traces after the facts.

I might also review all the path handling - it is likely an issue with this, although nothing obvious is visible in a procmon trace. Still I think we need not just fix this bug but also the general debugging infrastructure.

So it is making progress in a 4 steps forward, 3 steps back kind of walk.

**Comment by MSoegtropIMC** [comment]
I finally managed to track this down, a cut & paste error in gtksourceview:
```
BOOL WINAPI
DllMain (HINSTANCE hinstDLL,
	 DWORD     fdwReason,
	 LPVOID    lpvReserved)
{
	switch (fdwReason)
	{
		case DLL_PROCESS_ATTACH:
			gtksourceview_dll = hinstDLL;
			gtksourceview_init ();
			break;

		case DLL_THREAD_DETACH:
			gtksourceview_shutdown ();
			break;

		default:
			/* do nothing */
			break;
	}

	return TRUE;
}
```
So some data is killed whenever a thread ends. One just has to replace `DLL_THREAD_DETACH` with `DLL_PROCESS_DETACH` and voila ... This is already fixed in gtksourceview upstream, but not in the latest version available as Cygwin package. The reason might be that the build fails cause of some rather nasty things going on in gtksourceview autoconf. The Msys2 folks have a newer version of gtksourceview than cygwin - their trick is to skip the autoconf step and just run a configure file created on Linux.

The autoconf issues seems to exceed my M4 skills - maybe someone fluent in M4 can have a look during the workshop next week. Essentially what is happening is that some M4 macro is used in configure.ac and it is properly defined - it expands properly into configure - but there is a macro to check if the macro is defined and this check fails fails. Also a m4_dumpdef claims that the macro is undefined. Maybe this has to do with several phases of M4 processing. Interestingly the same mechanism is used for 10 different macros in configure.ac and it works for 8 and fails for 2. Also this seems to have worked a year back when the package was last compiled - I recompiled the package delivered with Cygwin. Maybe some regression in m4 or autoconf on Cygwin.

So you see I had some good sports in the last weeks, starting with fixing ocamldebug so that it can debug processes starting other OCaml processes (an issue which also exists on Linux) over debugging the interfaces between OCaml and C code by debugging a program in both ocamldebug and gdb at the same time to recompiling a lot of Cygwin MinGW packages because they lack proper debug infos to pondering over interesting convolutions of M4 macros in autconf.

But I think it was worthwhile - I have now a much better debug infrastructure, learned a lot about the internals of OCaml and am in the end more confident that I can fix other issues quickly in the future. In the end, after fixing all infrastructure issues, debugging the issue hardly takes 10 minutes.

And who would have thought that the final root cause of CoqIDE crashes is some M4 issues in autoconf ...

Now I need to get all the infrastructure fixes integrated upstream ...

**Comment by MSoegtropIMC** [comment]
I am pretty much happy now - last steps have been to select a reasonable subset of icons (the complete GTK3 set has a few 1000). CoqIDE seems to work with only minor oddities here and there now. And I still need to push a few things upstream.

There is one bad thing left though: If people have a file `C:\Users\<user>\AppData\Local\coq\coqiderc` from <=8.9.1 CoqIDE, syntax highlighting does not work unless one fixes a few settings (the language and highlighting style) in the preferences. I wonder how this is best handled. I would say deleting this file in the installer is not an option. One should rather make sure that this is reasonable compatible. I will have a quick look.

If the above file (and the .keys file in the same folder) is deleted, things work fine.

It might also be an option to give a note in the installer to delete this file in order to profit from new color schemes and so on. Things look quite different if I use a fixed coqiderc from GTK2 and if I delete it.
